packages/devtools/lib/src/timeline/timeline_protocol.dart

```diff
@@ -36,64 +60,133 @@ class TimelineData {
   ///
   /// Once frames are ready, we will remove them from this Map and add them to
   /// [_frameCompleteController].
-  final Map<String, TimelineFrame> _pendingFrames = <String, TimelineFrame>{};
+  final Map<String, TimelineFrame> pendingFrames = <String, TimelineFrame>{};
 
   /// Events we have collected and are waiting to add to their respective
   /// frames.
-  final List<TimelineEvent> _pendingEvents = [];
+  final List<TimelineEvent> pendingEvents = [];
 
-  /// The current node in the tree structure of CPU duration events.
-  TimelineEvent _cpuEventNode;
+  /// The current nodes in the tree structures of CPU and GPU duration events.
+  final List<TimelineEvent> currentEventNodes = List.generate(2, (_) => null);
 
-  /// The current node in the tree structure of GPU duration events.
-  TimelineEvent _gpuEventNode;
+  /// The previously handled DurationEnd event for both CPU and GPU.
+  ///
+  /// We need this information to balance the tree structures of our event nodes
+  /// if they fall out of balance due to duplicate trace events.
+  List<TraceEvent> previousDurationEndEvents = List.generate(2, (_) => null);
+
+  /// Heaps that order and store trace events as we receive them.
+  final List<HeapPriorityQueue<TraceEventWrapper>> heaps = List.generate(
+    2,
+    (_) => HeapPriorityQueue(TraceEventWrapper.traceComparator),
+  );
 
-  void processTimelineEvent(TraceEvent event) {
+  void processTraceEvent(TraceEvent event) {
     // TODO(kenzie): stop manually setting the type once we have that data from
     // the engine.
     event.type = _inferEventType(event);
 
-    if (!event.isGpuEvent && !event.isCpuEvent) return;
+    if (!_shouldProcessTraceEvent(event)) return;
 
-    if (_debugEventTrace) print(event.toString());
+    if (debugTimeline) {
+      debugTraceEvents.writeln(event.json.toString());
+    }
 
+    // Process flow events now. Process Duration events after a delay.
     switch (event.phase) {
       case 's':
-        _handleFrameStartEvent(event);
+        if (event.name.contains('PipelineItem')) {
```

Comment:
what should we be doing if this is not a PipelineItem?
document this special case here and elsewhere. Hopefully this special case can be removed as it does not seem robust.